_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+---
+title: "torch for optimization"
+description: > 
+  Torch is not just for deep learning. Its L-BFGS optimizer, complete with Strong-Wolfe line search, is a powerful tool in unconstrained as well as constrained optimization.
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchoptimization
+date: 04-27-2021
+categories:
+  - Torch
+  - R
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+So far, all `torch` use cases we've seen have been in deep learning. However, its automatic differentiation feature is useful in other areas, as well. One prominent example is numerical optimization: We can use `torch` to find the minimum of a function.
+
+In fact, function minimization is *exactly* what happens in training a neural network. But there, the function in question normally is far too complex to even imagine finding its minima analytically. Numerical optimization aims at building up the tools to handle just this complexity. To that end, however, it starts from functions that are far less deeply composed. Instead, they are hand-crafted to pose specific challenges.
```

Comment:
This paragraph is a little bit confusing to me, which I'm sure is at least in part due to my lack of technical understanding. Here's what I'm getting from it, and you can decide if any of this might be refactored:

* Function minimization is what happens when you're training a neural network, but that function is too complicated to imagine finding its minima.
* Numerical optimization is how we make tools to handle the unimaginably complex kind of function(?)
* To do this is starts with functions that are "less deeply composed" (so, I'm guessing easier to imagine? i.e. we can't imagine the ones from neural networks because they're super deeply composed?)

Summing up the last sentence but filling in the blanks/pronouns with what we've learned above:
* Numerical optimization starts from functions that are hand-crafted (by us?) to pose specific challenges (which are less complex, so we can imagine them)

Reply:
Yes, exactly! (One thing I wasn't hinting at was imagination; the problem with deep composition is more like that derivatives get super tedious to compute, so you wouldn't want do them end-to-end... instead what automatic differentiation libraries like torch do is building them up piece-by-piece, and just storing the intermediate updates to the end-to-end derivatives...)